### PR TITLE
Batch Norm Dynamic Batch size 1 and 2

### PR DESCRIPTION
**Pull request by daniellowell**
Dynamic implementations of batch size 1 and 2 forward training. Performance is better than develop. Could be improved with smaller workgroups for smaller images, but then performance suffer for larger images.

Tested on Radeon VII:

## Batch size 1:
  | PR | develop | ratio
-- | -- | -- | --
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 1 -c 1024 -H 14 -W 14   -m ${1} --forw 2 -s 1 -r 1 | 0.017 | 0.031 | 55%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 1 -c 128 -H 28 -W 28   -m ${1} --forw 2 -s 1 -r 1 | 0.009 | 0.013 | 69%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 1 -c 2048 -H 7 -W 7   -m ${1} --forw 2 -s 1 -r 1 | 0.029 | 0.056 | 52%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 1 -c 256 -H 14 -W 14   -m ${1} --forw 2 -s 1 -r 1 | 0.008 | 0.012 | 67%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 1 -c 256 -H 56 -W 56   -m ${1} --forw 2 -s 1 -r 1 | 0.028 | 0.031 | 88%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 1 -c 512 -H 28 -W 28   -m ${1} --forw 2 -s 1 -r 1 | 0.014 | 0.029 | 48%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 1 -c 512 -H 7 -W 7 -m   ${1} --forw 2 -s 1 -r 1 | 0.013 | 0.018 | 74%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 1 -c 64 -H 112 -W 112   -m ${1} --forw 2 -s 1 -r 1 | 0.030 | 0.040 | 75%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 1 -c 64 -H 56 -W 56   -m ${1} --forw 2 -s 1 -r 1 | 0.013 | 0.013 | 100%
  |   |   |  

## Batch size 2:
  | PR | develop | ratio
-- | -- | -- | --
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 2 -c 1024 -H 14 -W 14   -m ${1} --forw 2 -s 1 -r 1 | 0.021 | 0.031 | 69%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 2 -c 128 -H 28 -W 28   -m ${1} --forw 2 -s 1 -r 1 | 0.008 | 0.013 | 57%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 2 -c 2048 -H 7 -W 7   -m ${1} --forw 2 -s 1 -r 1 | 0.035 | 0.055 | 65%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 2 -c 256 -H 14 -W 14   -m ${1} --forw 2 -s 1 -r 1 | 0.009 | 0.011 | 80%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 2 -c 256 -H 56 -W 56   -m ${1} --forw 2 -s 1 -r 1 | 0.035 | 0.050 | 69%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 2 -c 512 -H 28 -W 28   -m ${1} --forw 2 -s 1 -r 1 | 0.017 | 0.030 | 56%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 2 -c 512 -H 7 -W 7 -m   ${1} --forw 2 -s 1 -r 1 | 0.013 | 0.017 | 81%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 2 -c 64 -H 112 -W 112   -m ${1} --forw 2 -s 1 -r 1 | 0.046 | 0.058 | 80%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 2 -c 64 -H 56 -W 56   -m ${1} --forw 2 -s 1 -r 1 | 0.016 | 0.019 | 86%
  |   |   |

### CODE REVIEW

File: src/ocl/batchnormocl.cpp [hunk 1]
```diff
@@ -160,15 +160,26 @@ void BatchNormForwardTraining(Handle& handle,
...
+        else if((in_nhw < 33554432 && in_cstride > 1024) || ((in_cstride > 60) && bfpmixparm))
         {
             //
         }
         else if(in_cstride <= 512)
         {
             variant = 0;
         }
-        else
+        else if(((n > 768) && (in_cstride > 150) && bfp32parm) || in_nhw >= 33554432)
```
Comment (muralinr) [inline src/ocl/batchnormocl.cpp:182]: Do we need in_nhw >= 33554432 here?  I don't see that in original code..

**Review by muralinr — APPROVED** [review]
Is there any performance impact on other configs?

File: src/ocl/batchnormocl.cpp (hunk 1 above)
Comment (daniellowell) [reply src/ocl/batchnormocl.cpp:182]: Compressed the logic. Please check if it amounts to the same as before. I’ll double check.

Comment (muralinr) [reply src/ocl/batchnormocl.cpp:182]: It seems fine to me.  If your performance testing is good,  Can you please merge it?

**Review by muralinr — APPROVED** [review]

File: src/ocl/batchnormocl.cpp [hunk 2]
```diff
@@ -160,7 +160,18 @@ void BatchNormForwardTraining(Handle& handle,
         unsigned int ldsgcn   = xlocalsize / 64;
         unsigned int ldsnogcn = xlocalsize;
         std::string algo_name = "miopenBatchNormForwardTrainingSpatial";
-        if((in_nhw < 33554432 && in_cstride > 1024) || ((in_cstride > 60) && bfpmixparm))
+
```
Comment (muralinr) [inline src/ocl/batchnormocl.cpp:163]: It's fine now..

**Review by muralinr — APPROVED** [review]

**Review by atamazov — COMMENTED** [review]
LGTM!
